sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala

```diff
@@ -79,12 +81,669 @@ object GpuHashJoin {
+   * If it is true variable width types will have a value returned that corresponds to a
+   * null, and unknown types will throw an exception.
+   */
+  private def calcRowSizeBits(dt: DataType, nullValueCalc: Boolean): Option[Int] = dt match {
+    case StructType(fields) =>
+      sumRowSizesBits(fields.map(_.dataType), nullValueCalc)
+    case dt: DecimalType if dt.precision > DType.DECIMAL64_MAX_PRECISION =>
+      if (nullValueCalc) {
+        throw new IllegalArgumentException(s"Found an unsupported type $dt")
+      } else {
+        None
+      }
+    case _: NumericType | DateType | TimestampType | BooleanType | NullType =>
+      Some(GpuColumnVector.getNonNestedRapidsType(dt).getSizeInBytes * 8 + 1)
+    case StringType | BinaryType | ArrayType(_, _) if nullValueCalc =>
+      // Single offset value and a validity value
+      Some((DType.INT32.getSizeInBytes * 8) + 1)
+    case x if nullValueCalc =>
+      throw new IllegalArgumentException(s"Found an unsupported type $x")
+    case _ => None
+  }
+}
+
+/**
+ * JoinGatherer for a single map/table
+ */
+class JoinGathererImpl(
+    // TODO need a way to spill/cache the GatherMap
+    private val gatherMap: GatherMap,
+    private val data: LazySpillableColumnarBatch,
+    private val closeData: Boolean) extends JoinGatherer {
+
+  // How much of the gather map we have output so far
+  private var gatheredUpTo: Long = 0
+  private val totalRows: Long = gatherMap.getRowCount
+  private val totalInputRows: Int = data.numRows
+  private val totalInputSize: Long = data.deviceMemorySize
+  private val (fixedWidthRowSizeBits, nullRowSizeBits) = {
+    val dts = data.dataTypes
+    val fw = JoinGathererImpl.fixedWidthRowSizeBits(dts)
+    val nullVal = JoinGathererImpl.nullRowSizeBits(dts)
+    (fw, nullVal)
+  }
+
+  override def realCheapPerRowSizeEstimate: Double = {
+    // Avoid divide by 0 here and later on
+    if (totalInputRows > 0 && totalInputSize > 0) {
+      totalInputSize.toDouble / totalInputRows
+    } else {
+      1.0
+    }
+  }
+
+  override def getFixedWidthBitSize: Option[Int] = fixedWidthRowSizeBits
+
+  override def gatherNext(n: Int): ColumnarBatch = synchronized {
+    val start = gatheredUpTo
+    assert((start + n) <= totalRows)
+    val ret = withResource(gatherMap.toColumnView(start, n)) { gatherView =>
+      val batch = data.getBatch
+      val gatheredTab = withResource(GpuColumnVector.from(batch)) { table =>
+        table.gather(gatherView)
+      }
+      withResource(gatheredTab) { gt =>
+        GpuColumnVector.from(gt, GpuColumnVector.extractTypes(batch))
+      }
+    }
+    gatheredUpTo += n
+    ret
+  }
+
+  override def isDone: Boolean = synchronized {
+    gatheredUpTo >= totalRows
+  }
+
+  override def numRowsLeft: Long = totalRows - gatheredUpTo
+
+  override def allowSpilling(): Unit = {
+    data.allowSpilling()
+  }
+
+  override def getBitSizeMap(n: Int): ColumnView = synchronized {
```

Comment:
Feels like this function could use a function header comment. As I understand it, it computes bit size per row of a cudf Table and returns a single column of longs with those values.